e2e: sub package refactoring

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The goal of https://github.com/kubernetes/kubernetes/issues/81232 and https://github.com/kubernetes/kubernetes/issues/81245 (duplicates?) was to refactor the framework so that test/e2e/framework is smaller, with less dependency on features and packages that might not be needed. The intention was to then publish it for external consumption.

That is no longer a goal. However, refactoring the framework like that still has benefits:
- cleaner architecture
- API easier to read and discover

That makes it worthwhile to finish the original effort.

Another motivation is that the original effort got stuck halfway through and left the framework in limbo: the actual dependencies became the opposite (framework depends on sub packages) of what they should be (sub packages depend on framework). This made it impossible to add new core functionality that can be used by both core and sub packages. We either would have to revert or finish the work.

#### Which issue(s) this PR fixes:

Fixes #https://github.com/kubernetes/kubernetes/issues/81245

#### Special notes for your reviewer:

The approach taken with this PR was to move code incrementally out of core framework. Putting that code into the existing sub packages would have immediately led to import cycles, so instead test/e2e/framework/todo was created with sub packages that mirror the "real" ones. Those "todo" sub packages may depend on the core framework without breaking the compilation.

As a result, `go build ./test/e2e/framework/...` always passes for each commit. However, other code was not updated, so the full e2e test suite doesn't build (yet).

Once dependencies were as expected, the code under "todo" was moved into the real sub packages in the final commit.

#### Does this PR introduce a user-facing change?

**If you have code that no longer compiles because of this PR, you can use the script from the commit message in https://github.com/kubernetes/kubernetes/pull/112043/commits/dfdf88d4faafa6fd39988832ea0ef6d668f490e9 to update that code.**

```release-note
The test/e2e/framework was refactored so that the core framework is smaller. Optional functionality like resource monitoring, log size monitoring, metrics gathering and debug information dumping must be imported by specific e2e test suites. Init packages are provided which can be imported to re-enable the functionality that traditionally was in the core framework. If you have code that no longer compiles because of this PR, you can use the script [from a commit message](https://github.com/kubernetes/kubernetes/pull/112043/commits/dfdf88d4faafa6fd39988832ea0ef6d668f490e9) to update that code.
```

pohly committed:
e2e framework: move kubectl and pod helper code

pohly committed:
e2e framework: move metrics gathering into sub package

This reduces the size of the test/e2e/framework itself. Because it does not
gather metrics data anymore by default, E2E test suites must set their
callbacks function or set the original one by importing
"k8s.io/kubernetes/test/e2e/framework/todo/metrics/init".

pohly committed:
e2e framework: move node helper code into sub package

This reduces the size of the test/e2e/framework itself. Because it does not
check nodes anymore by default, E2E test suites must set their own check
function or set the original one by importing
"k8s.io/kubernetes/test/e2e/framework/todo/node/init".

pohly committed:
e2e framework: move dumping of information into sub package

This reduces the size of the test/e2e/framework itself. Because it does not
dump anything anymore by default, E2E test suites must set their own dump
function or set the original one by importing
"k8s.io/kubernetes/test/e2e/framework/debug/init".

pohly committed:
e2e framework: support callbacks during framework creation and before each test

This will be used as mechanism for invoking some of the code which is currently
hard-coded in the framework once that code is placed in optional packages.

pohly committed:
e2e framework: move resource gathering into framework/debug

This helps getting rid of the ssh dependency. The same init package as for
dumping namespaces takes care of adding the functionality back to framework
instances.

pohly committed:
e2e framework: move log size verification into framework/debug

This helps getting rid of the ssh dependency. The same init package as for
dumping namespaces takes care of adding the functionality back to framework
instances.

pohly committed:
e2e framework: move NodeKiller into framework/node

This removes some more dependencies from the framework.

Removing the configuration needs to be handled as part of refactoring the
TestContext.

pohly committed:
e2e framework: move pod exec and create into sub package

pohly committed:
e2e framework: eliminate redundant framework/[log|ginkgowrapper]

These sub packages were created by mistake. Logging and failure handling are
core features and must be implemented in the framework package.

pohly committed:
e2e framework: eliminate interim sub packages

The "todo" packages were necessary while moving code around to avoid hitting
cyclic dependencies. Now that any sub package can depend on the framework, they
are no longer needed and the code can be moved into the normal sub packages.

pohly committed:
e2e framework: introduce todo sub packages

aojea commented:
/assign @aojea @spiffxp 

Aaron is probably the person with more historical context, we'll need to have this context :smile:

pohly commented:
I already discussed it with him on Slack: https://kubernetes.slack.com/archives/C09QZ4DQB/p1661356920849319

pohly committed:
e2e: make import blocks consistent

pohly committed:
e2e: adapt to moved code

This is the result of automatically editing source files like this:

    go install golang.org/x/tools/cmd/goimports@latest
    find ./test/e2e* -name "*.go" | xargs env PATH=$GOPATH/bin:$PATH ./e2e-framework-sed.sh

with e2e-framework-sed.sh containing this:

sed -i \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.ExecCommandInContainer(/e2epod.ExecCommandInContainer(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.ExecCommandInContainerWithFullOutput(/e2epod.ExecCommandInContainerWithFullOutput(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.ExecShellInContainer(/e2epod.ExecShellInContainer(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.ExecShellInPod(/e2epod.ExecShellInPod(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.ExecShellInPodWithFullOutput(/e2epod.ExecShellInPodWithFullOutput(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.ExecWithOptions(/e2epod.ExecWithOptions(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.MatchContainerOutput(/e2eoutput.MatchContainerOutput(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.PodClient(/e2epod.NewPodClient(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.PodClientNS(/e2epod.PodClientNS(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.TestContainerOutput(/e2eoutput.TestContainerOutput(\1, /" \
    -e "s/\(f\|fr\|\w\w*\.[fF]\w*\)\.TestContainerOutputRegexp(/e2eoutput.TestContainerOutputRegexp(\1, /" \
    -e "s/framework.AddOrUpdateLabelOnNode\b/e2enode.AddOrUpdateLabelOnNode/" \
    -e "s/framework.AllNodes\b/e2edebug.AllNodes/" \
    -e "s/framework.AllNodesReady\b/e2enode.AllNodesReady/" \
    -e "s/framework.ContainerResourceGatherer\b/e2edebug.ContainerResourceGatherer/" \
    -e "s/framework.ContainerResourceUsage\b/e2edebug.ContainerResourceUsage/" \
    -e "s/framework.CreateEmptyFileOnPod\b/e2eoutput.CreateEmptyFileOnPod/" \
    -e "s/framework.DefaultPodDeletionTimeout\b/e2epod.DefaultPodDeletionTimeout/" \
    -e "s/framework.DumpAllNamespaceInfo\b/e2edebug.DumpAllNamespaceInfo/" \
    -e "s/framework.DumpDebugInfo\b/e2eoutput.DumpDebugInfo/" \
    -e "s/framework.DumpNodeDebugInfo\b/e2edebug.DumpNodeDebugInfo/" \
    -e "s/framework.EtcdUpgrade\b/e2eproviders.EtcdUpgrade/" \
    -e "s/framework.EventsLister\b/e2edebug.EventsLister/" \
    -e "s/framework.ExecOptions\b/e2epod.ExecOptions/" \
    -e "s/framework.ExpectNodeHasLabel\b/e2enode.ExpectNodeHasLabel/" \
    -e "s/framework.ExpectNodeHasTaint\b/e2enode.ExpectNodeHasTaint/" \
    -e "s/framework.GCEUpgradeScript\b/e2eproviders.GCEUpgradeScript/" \
    -e "s/framework.ImagePrePullList\b/e2epod.ImagePrePullList/" \
    -e "s/framework.KubectlBuilder\b/e2ekubectl.KubectlBuilder/" \
    -e "s/framework.LocationParamGKE\b/e2eproviders.LocationParamGKE/" \
    -e "s/framework.LogSizeDataTimeseries\b/e2edebug.LogSizeDataTimeseries/" \
    -e "s/framework.LogSizeGatherer\b/e2edebug.LogSizeGatherer/" \
    -e "s/framework.LogsSizeData\b/e2edebug.LogsSizeData/" \
    -e "s/framework.LogsSizeDataSummary\b/e2edebug.LogsSizeDataSummary/" \
    -e "s/framework.LogsSizeVerifier\b/e2edebug.LogsSizeVerifier/" \
    -e "s/framework.LookForStringInLog\b/e2eoutput.LookForStringInLog/" \
    -e "s/framework.LookForStringInPodExec\b/e2eoutput.LookForStringInPodExec/" \
    -e "s/framework.LookForStringInPodExecToContainer\b/e2eoutput.LookForStringInPodExecToContainer/" \
    -e "s/framework.MasterAndDNSNodes\b/e2edebug.MasterAndDNSNodes/" \
    -e "s/framework.MasterNodes\b/e2edebug.MasterNodes/" \
    -e "s/framework.MasterUpgradeGKE\b/e2eproviders.MasterUpgradeGKE/" \
    -e "s/framework.NewKubectlCommand\b/e2ekubectl.NewKubectlCommand/" \
    -e "s/framework.NewLogsVerifier\b/e2edebug.NewLogsVerifier/" \
    -e "s/framework.NewNodeKiller\b/e2enode.NewNodeKiller/" \
    -e "s/framework.NewResourceUsageGatherer\b/e2edebug.NewResourceUsageGatherer/" \
    -e "s/framework.NodeHasTaint\b/e2enode.NodeHasTaint/" \
    -e "s/framework.NodeKiller\b/e2enode.NodeKiller/" \
    -e "s/framework.NodesSet\b/e2edebug.NodesSet/" \
    -e "s/framework.PodClient\b/e2epod.PodClient/" \
    -e "s/framework.RemoveLabelOffNode\b/e2enode.RemoveLabelOffNode/" \
    -e "s/framework.ResourceConstraint\b/e2edebug.ResourceConstraint/" \
    -e "s/framework.ResourceGathererOptions\b/e2edebug.ResourceGathererOptions/" \
    -e "s/framework.ResourceUsagePerContainer\b/e2edebug.ResourceUsagePerContainer/" \
    -e "s/framework.ResourceUsageSummary\b/e2edebug.ResourceUsageSummary/" \
    -e "s/framework.RunHostCmd\b/e2eoutput.RunHostCmd/" \
    -e "s/framework.RunHostCmdOrDie\b/e2eoutput.RunHostCmdOrDie/" \
    -e "s/framework.RunHostCmdWithFullOutput\b/e2eoutput.RunHostCmdWithFullOutput/" \
    -e "s/framework.RunHostCmdWithRetries\b/e2eoutput.RunHostCmdWithRetries/" \
    -e "s/framework.RunKubectl\b/e2ekubectl.RunKubectl/" \
    -e "s/framework.RunKubectlInput\b/e2ekubectl.RunKubectlInput/" \
    -e "s/framework.RunKubectlOrDie\b/e2ekubectl.RunKubectlOrDie/" \
    -e "s/framework.RunKubectlOrDieInput\b/e2ekubectl.RunKubectlOrDieInput/" \
    -e "s/framework.RunKubectlWithFullOutput\b/e2ekubectl.RunKubectlWithFullOutput/" \
    -e "s/framework.RunKubemciCmd\b/e2ekubectl.RunKubemciCmd/" \
    -e "s/framework.RunKubemciWithKubeconfig\b/e2ekubectl.RunKubemciWithKubeconfig/" \
    -e "s/framework.SingleContainerSummary\b/e2edebug.SingleContainerSummary/" \
    -e "s/framework.SingleLogSummary\b/e2edebug.SingleLogSummary/" \
    -e "s/framework.TimestampedSize\b/e2edebug.TimestampedSize/" \
    -e "s/framework.WaitForAllNodesSchedulable\b/e2enode.WaitForAllNodesSchedulable/" \
    -e "s/framework.WaitForSSHTunnels\b/e2enode.WaitForSSHTunnels/" \
    -e "s/framework.WorkItem\b/e2edebug.WorkItem/" \
    "$@"

for i in "$@"; do
    # Import all sub packages and let goimports figure out which of those
    # are redundant (= already imported) or not needed.
    sed -i -e '/"k8s.io.kubernetes.test.e2e.framework"/a e2edebug "k8s.io/kubernetes/test/e2e/framework/debug"' "$i"
    sed -i -e '/"k8s.io.kubernetes.test.e2e.framework"/a e2ekubectl "k8s.io/kubernetes/test/e2e/framework/kubectl"' "$i"
    sed -i -e '/"k8s.io.kubernetes.test.e2e.framework"/a e2enode "k8s.io/kubernetes/test/e2e/framework/node"' "$i"
    sed -i -e '/"k8s.io.kubernetes.test.e2e.framework"/a e2eoutput "k8s.io/kubernetes/test/e2e/framework/pod/output"' "$i"
    sed -i -e '/"k8s.io.kubernetes.test.e2e.framework"/a e2epod "k8s.io/kubernetes/test/e2e/framework/pod"' "$i"
    sed -i -e '/"k8s.io.kubernetes.test.e2e.framework"/a e2eproviders "k8s.io/kubernetes/test/e2e/framework/providers"' "$i"
    goimports -w "$i"
done

pohly committed:
build: adapt to new location of etcdImage variable

zeitgeist failed because it expected the variable in the old file.

leilajal commented:
/triage accepted

pohly commented:
This is no longer WIP: I went ahead and fixed all imports, plus adapted some paths elsewhere.

However, https://github.com/kubernetes/kubernetes/pull/111998 should get merged first because this PR is based on it.

michelle192837 commented:
/assign

pohly commented:
This PR currently includes changes at the beginning that were needed but have been merged in some slightly different form through different PRs.

Only the commits starting with "[e2e framework: support callbacks during framework creation and before…](https://github.com/kubernetes/kubernetes/pull/112043/commits/ffc305ca2743e870d70c97b39edfbbebe6e6cc52)" need to be reviewed.

I can rebase once it is certain that the PR then will be reviewed relatively soon.

pohly commented:
Actually, rebasing wasn't too bad. I've done it now.

marosset commented:
/test pull-kubernetes-e2e-capz-windows-containerd

spiffxp reviewed (commented):
I know this will get caught in rebase if it hangs out too long. I think this is mostly a move in the right direction, but I'm wary of immediately slapping approve on this without hearing feedback from other consumers/maintainers of the e2e framework.

I think my high level concerns are:
- The design intent / roadmap is AFAIK still undocumented... what is the end goal, and have you reached it here?
- The `NewFrameworkExtensions` mechanism introduces more complexity, is it really worth it?

spiffxp commented:
(btw thank you for breaking this up into commits the way you did, that really helped make this more digestible)

pohly commented on test/e2e/framework/todo/node/init/init.go:
```diff
@@ -0,0 +1,37 @@
+			})
+		},
+	)
+}
```
> The design intent / roadmap is AFAIK still undocumented... what is the end goal, and have you reached it here?

The goal is to have a smaller core framework with optional functionality moved into sub packages that a specific E2E suite may use, but doesn't have to import when not using it. I can add that in a README.md. This PR achieves that goal.

> One of the chief complaints about framework is that it's hard to know the order in which things are getting called. 

I suppose it didn't help either that there were various custom hooks and callbacks, with poor to no documentation... I agree that this extension mechanism adds some complexity, but I compensated for it a bit by removing some of those other custom hooks in https://github.com/kubernetes/kubernetes/pull/111998.

I can explain this all better in the top-level README that I'll add.

> This, to me, complicates things even further:
>
> -   the order in which these inits get imported matters

Which shouldn't matter. Each of them must be (and is) self-contained and the code that they execute during a test run runs in separate Ginkgo nodes, so even if one fails the others still execute.

>  -  the order in which the extracted blocks of code run looks, to my eyes, to be subtly different than before

There might be small changes, but I don't think they matter.

> - framework is effectively "incomplete" without these inits (or at the least its behavior without these is definitely something we don't have experience with)

True, downstream users who expect the same behavior as before need to make changes. But such downstream users are using k/k at their own risk, so this shouldn't hold us back. All of the three in-tree e2e suites behave as before.

I am one of the downstream users who decided that this risk of random breakages is worth it, and with that hat on I can say that this change seems useful and manageable.

> My understanding is you did this to break parts of framework.BeforeEach and framework.AfterEach into sub packages, though I'm less clear on why it was necessary.

If I hadn't done that, the core framework would have had a hard dependency on several of the sub packages. Using this example here, several parts of `test/e2e/framework/node` would have to be moved back into `test/e2e/framework` to avoid the `framework -> node` dependency that exists prior to this PR.

> Are you sure there's no other way of accomplishing what you're trying to do here?

I can't think of one, short of merging everything back into a single monolithic package (but that's not the goal).

What we need is a way for a test suite author to inject new functionality into each test, but without having to to modify every single test. Each test calls `framework.NewDefaultFramework`, so an extension mechanism for that seemed suitable.

pohly commented on test/e2e/framework/auth/helpers.go:
```diff
@@ -29,7 +29,7 @@ import (
 	v1authorization "k8s.io/client-go/kubernetes/typed/authorization/v1"
 	v1rbac "k8s.io/client-go/kubernetes/typed/rbac/v1"
-	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+	"k8s.io/kubernetes/test/e2e/framework"
```
> "Logging and failure handling are core features and must be implemented in the framework package."

This came from the original developers who tried (unsuccessfully) to move that functionality into sub packages. The reason why it cannot work is that the framework itself must be able to log certain things and has error scenarios where it must call the fail handler.

I'll try to describe the rationale in the upcoming top-level README.

pohly replied:
README.md added.

pohly commented on test/e2e/framework/debug/init/init.go:
```diff
@@ -28,6 +32,31 @@ func init() {
+					closeChannel = make(chan bool)
+					verifier = debug.NewLogsVerifier(f.ClientSet, closeChannel)
+					go func() {
+						defer wg.Done()
```
It's the normal pattern for `wg.Done` and has the advantage that it gets called also when some code in the function panics.

The original code was:
```
	go func() {
			f.logsSizeVerifier.Run()
			f.logsSizeWaitGroup.Done()
		}()
```

If `Run` paniced, `Done` wouldn't get called.

This seemed like a useful "while I am touching the code" improvement.

pohly commented on test/e2e/framework/debug/init/init.go:
```diff
@@ -57,6 +69,42 @@ func init() {
+					ginkgo.DeferCleanup(func() {
+						ginkgo.By("Collecting resource usage data", func() {
+							summary, resourceViolationError := gatherer.StopAndSummarize([]int{90, 99, 100}, AddonResourceConstraints)
+							framework.ExpectNoError(resourceViolationError)
```
Original code:
```
summary, resourceViolationError := f.gatherer.StopAndSummarize([]int{90, 99, 100}, f.AddonResourceConstraints)
defer ExpectNoError(resourceViolationError)
f.TestSummaries = append(f.TestSummaries, summary)
```

This looked like a bug to me - why use `summary` when there was an error? But you are right, this is a change of behavior.

I'll restore the original order, albeit with a bit simpler code and a comment:
```
summary, resourceViolationError := gatherer.StopAndSummarize([]int{90, 99, 100}, AddonResourceConstraints)
// Always use the result, even when there was an error.
f.TestSummaries = append(f.TestSummaries, summary)			
// Now fail if there was an error.
framework.ExpectNoError(resourceViolationError)
```

pohly replied:
Done in the updated commit.

pohly commented on test/e2e/framework/debug/init/init.go:
```diff
@@ -15,35 +15,47 @@ limitations under the License.
+)
+
+var (
+	// TODO: this variable used to be a field in framework.Framework. It is
+	// not clear how it was ever set. https://grep.app/search?q=AddonResourceConstraints
+	// returns only the default initialization with an empty map. Perhaps it can be removed?
```
Or do it right away in this PR, in a separate commit?

pohly replied:
I've added one commit for it.

pohly committed:
e2e framework: also test BeforeEach and extension mechanism

This adds test coverage for NewFrameworkExtensions and shows better how
BeforeEach callbacks are invokved. The unit test is not strictly about just the
cleanup operations anymore, but that's okay(ish).

pohly committed:
e2e framework: add top-level README.md

This documents the cleaned up package structure and explains how callbacks are
invoked. Ginkgko v2 has led to some improvements in that area.

pohly committed:
e2e framework: remove AddonResourceConstraints

According to spiffx in
https://github.com/kubernetes/kubernetes/pull/112043/files#r978031680:

> Last in-tree tests that used it were removed in #83322 (1.17 timeframe)
> so I think it can be safely removed

pohly commented:
> I know this will get caught in rebase if it hangs out too long.

Yes, I had to rebase again today :sweat_smile:  This makes a comparison against the revision that you reviewed harder, too. I changed one commit ([e2e framework: move resource gathering into framework/debug](https://github.com/kubernetes/kubernetes/pull/112043/commits/0b23b4e56bce1d51722ee8fc7163043f18a1bbd3)) and add three new commits at the end.

> I think this is mostly a move in the right direction, but I'm wary of immediately slapping approve on this without hearing feedback from other consumers/maintainers of the e2e framework.

I asked on #sig-testing for reviewers (not just this PR, but in general), but it looks like only @aojea and myself are active at the moment. I am also a consumer of the E2E framework outside of Kubernetes.

@mythi: you are also using the E2E framework in several projects. What do you think about this change?

@aojea: any comments?

pohly commented:
Moving all of these functions around will break downstream test suites. It's worth calling out that the script from the commit message in https://github.com/kubernetes/kubernetes/pull/112043/commits/9c755e01db11a31f40283c221627e8afd9ea505c can be used to automatically fix up also downstream tests.

I was tempted to put that into a script inside the repo, but where, and how long should we keep it there?

pohly commented:
/retest

pohly commented:
/retest

mythi commented:
> > I think this is mostly a move in the right direction, but I'm wary of immediately slapping approve on this without hearing feedback from other consumers/maintainers of the e2e framework.
> 
> I asked on #sig-testing for reviewers (not just this PR, but in general), but it looks like only @aojea and myself are active at the moment. I am also a consumer of the E2E framework outside of Kubernetes.
> 
> @mythi: you are also using the E2E framework in several projects. What do you think about this change?

Yes, we are pulling the k8s e2e framework run our automated device plugins tests. This PR make a lot of sense to me. Quoting the README:

> The advantages of splitting the code like this are:
> - leaner go doc packages by grouping related functions together

as an external user I often need to look for functionality xyz (e.g., get pod logs, wait for pods) in framework code and I find that cumbersome. This proposal seems to help with that by moving things in more logical places.

> - not forcing all E2E suites to import all functionality

I see this as big benefit (although it remains to be seen how much it really helps to reduce the dependencies in our e2e suite).

spiffxp reviewed (approved):
/approve
/lgtm
I guess I would rather see this land before the volume of code changes starts to pick up.

One last concern (which I don't really know how to avoid so we'll find out) is that cherry-picks that touch e2e tests will likely require some manual changes to undo the renames/migrations.  

Taking the script in the commit message in https://github.com/kubernetes/kubernetes/pull/112043/commits/9c755e01db11a31f40283c221627e8afd9ea505c and reversing it would be a good starting point if someone wants to automatically undo them.  If it turns out to be a repeated pain, we could put such a thing in `hack/` but I'm fine waiting to see if it's necessary.

spiffxp commented:
```
test/e2e/kubectl/kubectl.go:1948:14: undefined: framework.RunKubectlOrDie
test/e2e/kubectl/kubectl.go:1958:24: undefined: framework.RunKubectlOrDie
test/e2e/kubectl/kubectl.go:1965:23: undefined: framework.RunKubectlOrDie
```

Looks like this will need another rebase

pohly commented:
Rebased - quick, someone please LGTM it while it is still fresh :sweat_smile:

aojea commented:
/lgtm

pohly commented:
/retest

spiffxp commented:
Just to give PR authors a heads up in case this caused a merge-conflict, this is a naive "do they touch the same files" list of open PR's that might need to rename / import a few things (see https://github.com/kubernetes/kubernetes/commit/9c755e01db11a31f40283c221627e8afd9ea505c):
- https://github.com/kubernetes/kubernetes/pull/86139
- https://github.com/kubernetes/kubernetes/pull/98307
- https://github.com/kubernetes/kubernetes/pull/100335
- https://github.com/kubernetes/kubernetes/pull/102025
- https://github.com/kubernetes/kubernetes/pull/107514
- https://github.com/kubernetes/kubernetes/pull/107760
- https://github.com/kubernetes/kubernetes/pull/107958
- https://github.com/kubernetes/kubernetes/pull/109924
- https://github.com/kubernetes/kubernetes/pull/109944
- https://github.com/kubernetes/kubernetes/pull/110159
- https://github.com/kubernetes/kubernetes/pull/110318
- https://github.com/kubernetes/kubernetes/pull/110863
- https://github.com/kubernetes/kubernetes/pull/110991
- https://github.com/kubernetes/kubernetes/pull/110994
- https://github.com/kubernetes/kubernetes/pull/111124
- https://github.com/kubernetes/kubernetes/pull/111184
- https://github.com/kubernetes/kubernetes/pull/111204
- https://github.com/kubernetes/kubernetes/pull/111211
- https://github.com/kubernetes/kubernetes/pull/111552
- https://github.com/kubernetes/kubernetes/pull/111735
- https://github.com/kubernetes/kubernetes/pull/111905
- https://github.com/kubernetes/kubernetes/pull/111930
- https://github.com/kubernetes/kubernetes/pull/111997
- https://github.com/kubernetes/kubernetes/pull/112019
- https://github.com/kubernetes/kubernetes/pull/112255
- https://github.com/kubernetes/kubernetes/pull/112546

spiffxp commented:
I think this is causing https://github.com/kubernetes/kubernetes/issues/112996
